kazoo/recipe/cache.py

```diff
@@ -0,0 +1,382 @@
+                        does not exist.
+        :raises ValueError: If the path is outside of this subtree.
+        :returns: The :class:`frozenset` which including children names.
+        """
+        node = self._find_node(path)
+        return default if node is None else frozenset(node._children)
+
+    def _find_node(self, path):
+        if not path.startswith(self._root._path):
+            raise ValueError('outside of tree')
+        striped_path = path[len(self._root._path):].strip('/')
+        splited_path = [p for p in striped_path.split('/') if p]
+        current_node = self._root
+        for node_name in splited_path:
+            if node_name not in current_node._children:
+                return
+            current_node = current_node._children[node_name]
+        return current_node
+
+    def _publish_event(self, event_type, event_data=None):
+        event = TreeEvent.make(event_type, event_data)
+        if self._state != self.STATE_CLOSED:
+            logger.debug('public event: {0!r}'.format(event))
+            self._in_background(self._do_publish_event, event)
+
+    def _do_publish_event(self, event):
+        for listener in self._event_listeners:
+            with handle_exception(self._error_listeners):
+                listener(event)
+
+    def _in_background(self, func, *args, **kwargs):
+        self._client.handler.callback_queue.put(lambda: func(*args, **kwargs))
+
+    def _session_watcher(self, state):
+        if state == KazooState.SUSPENDED:
+            self._publish_event(TreeEvent.CONNECTION_SUSPENDED)
+        elif state == KazooState.CONNECTED:
+            with handle_exception(self._error_listeners):
+                self._root.was_reconnected()
+                self._publish_event(TreeEvent.CONNECTION_RECONNECTED)
+        elif state == KazooState.LOST:
+            self._is_initialized = False
+            self._publish_event(TreeEvent.CONNECTION_LOST)
+
+
+class TreeNode(object):
+    """The tree node record.
+
+    :param tree: A :class:`~kazoo.recipe.cache.TreeCache` instance.
+    :param path: The path of current node.
+    :param parent: The parent node reference. ``None`` for root node.
+    """
+
+    __slots__ = ('_tree', '_path', '_parent', '_depth', '_children', '_state',
+                 '_data')
+
+    STATE_PENDING = 0
+    STATE_LIVE = 1
+    STATE_DEAD = 2
+
+    def __init__(self, tree, path, parent):
+        self._tree = tree
+        self._path = path
+        self._parent = parent
+        self._depth = parent._depth + 1 if parent else 0
+        self._children = {}
+        self._state = self.STATE_PENDING
+        self._data = None
+
+    @classmethod
+    def make_root(cls, tree, path):
+        return cls(tree, path, None)
+
+    def was_reconnected(self):
+        self._refresh()
+        for child in self._children.values():
+            child.was_reconnected()
+
+    def was_created(self):
+        self._refresh()
+
+    def was_deleted(self):
```

Comment:
@harlowja I believe it was ported from [the original code base](https://git-wip-us.apache.org/repos/asf?p=curator.git;a=blob;f=curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/TreeCache.java;h=ed32223819fb5d78004142c26f621e7eb6b3052b;hb=refs/heads/CURATOR-3.0#l287).